### PR TITLE
Wdlupdate

### PR DESCRIPTION
**Pull request by bshifaw**
pairedtoubam Wdl now uses a readgroup tsv file as input, also added task to compose a file containing a list of the generated ubams

### CODE REVIEW

File: paired-fastq-to-unmapped-bam.inputs.json [hunk 1]
```diff
@@ -1,40 +1,22 @@
...
-  "ConvertPairedFastQsToUnmappedBamWf.PairedFastQsToUnmappedBAM.docker": "broadinstitute/gatk:4.0.1.2",
-  "ConvertPairedFastQsToUnmappedBamWf.PairedFastQsToUnmappedBAM.mem_size": "1 GB",
-  "ConvertPairedFastQsToUnmappedBamWf.PairedFastQsToUnmappedBAM.disk_size": 200
-}
+  "##_Comment3": "Docker",
+  "ConvertPairedFastQsToUnmappedBamWf.docker": "broadinstitute/gatk:4.0.2.0",
+
+  "##_Comment3": "Disk Space",
+  "##_ConvertPairedFastQsToUnmappedBamWf.PairedFastQsToUnmappedBAM.disk_space_gb": "(optional) Int?",
+  "##_ConvertPairedFastQsToUnmappedBamWf.PairedFastQsToUnmappedBAM.machine_mem_gb": "(optional) Int?",
```
Comment (ruchim) [inline paired-fastq-to-unmapped-bam.inputs.json]: Listed under disk space but this is a memory parameter

File: paired-fastq-to-unmapped-bam.inputs.json [hunk 2]
```diff
@@ -1,40 +1,22 @@
...
 
+  "##_Comment3": "Memory",
+  "##_ConvertPairedFastQsToUnmappedBamWf.CreateUbamList.disk_space_gb": "(optional) Int?",
```
Comment (ruchim) [inline paired-fastq-to-unmapped-bam.inputs.json]: Listed under memory but this is a disk space parameter

File: paired-fastq-to-unmapped-bam.wdl [hunk 3]
```diff
@@ -26,31 +29,46 @@
...
+  File readgroup_list  
+  Array[Array[String]] readgroup_array = read_tsv(readgroup_list) 
+  String ubam_list_name = basename(readgroup_list,".list")
+  
+  String docker
+  Int? preemptible_attempts
 
   # Convert multiple pairs of input fastqs in parallel
-  scatter (readgroup in readgroup_list) {
+  scatter (i in range(length(readgroup_array))) {
```
Comment (ruchim) [inline paired-fastq-to-unmapped-bam.wdl]: I like that the original scatter syntax is less verbose:

```
scatter (readgroup in readgroup_list) {
 
  call PairedFastQsToUnmappedBAM {
    input:
      fastq_1 = readgroup_array[1],
      fastq_2 = readgroup_array[2],
      ...
    }
}
```

File: paired-fastq-to-unmapped-bam.wdl [hunk 4]
```diff
@@ -26,31 +29,46 @@
...
+        preemptible_attempts = preemptible_attempts
+    }
+   }
+
+    #Create a list with the generated ubams
+    call CreateUbamList {
+      input:
+        unmapped_bams = PairedFastQsToUnmappedBAM.output_bam,
+        ubam_list_name = ubam_list_name,
+	    docker = docker,
```
Comment (ruchim) [inline paired-fastq-to-unmapped-bam.wdl]: adjusting spacing

File: paired-fastq-to-unmapped-bam.wdl [hunk 5]
```diff
@@ -88,12 +108,38 @@ task PairedFastQsToUnmappedBAM {
...
-    disks: "local-disk " + disk_size + " HDD"
+    disks: "local-disk " + select_first([disk_space_gb, 100]) + " HDD"
+    preemptible: select_first([preemptible_attempts, 3])
   }
   output {
     File output_bam = "${readgroup_name}.unmapped.bam"
   }
 }
 
+task CreateUbamList {
```
Comment (ruchim) [inline paired-fastq-to-unmapped-bam.wdl]: Can this task be simplified by using a write_lines()

```
command {
 mv ${write_lines(unmapped_bams)}  ${ubam_list_name}.unmapped_bams.list
}
```

I also don't think it needs to explicitly specify disk/mem as the default Cromwell values should be sufficient. You could probably get away with just docker & preemptible count, or even hard-code preemptible to 3 as such a short task will likely not need more than one attempt.

```
runtime {
   docker: docker
   preemptible: select_first([preemptible_attempts, 3])
}
```

File: README.md [hunk 6]
```diff
@@ -23,9 +23,14 @@ This WDL converts paired FASTQ to uBAM and adds read group information
 #### Requirements/expectations
 - Pair-end sequencing data in FASTQ format (one file per orientation)
 - One or more read groups, one per pair of FASTQ files 
+- The wdl will accept a file containing a list of the readgroups along with their metadata as the input. The wdl expects a specifi format for the colomuns, please follow the format as provided :
```
Comment (ruchim) [inline README.md]: It's not important but maybe reword this sentence a bit? 

This WDL requires an input TSV file such that each row corresponds to a readgroup and the columns contain metadata for each readgroup. The WDL depends on the input TSV to follow a specific format for the columns. Please use the template provided below as the header for the input TSV file:

**Review by ruchim — CHANGES_REQUESTED** [review]
Looks good overall, just some suggestions to simplify the workflow/task definition.

File: paired-fastq-to-unmapped-bam.wdl (hunk 3 above)
Comment (bshifaw) [reply paired-fastq-to-unmapped-bam.wdl]: We switched to using a readgroup list file instead of an array because there isn't an option for map variables in FC for the other input variables (fastq_pairs and metadata)

File: bam-to-unmapped-bams.wdl [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+## This WDL converts BAM  to unmapped BAMs
+##
+## Requirements/expectations :
+## - BAM file
+##
+## Outputs :
+## - Sorted Unmapped BAMs
+##
+## Cromwell version support
+## - Successfully tested on v31
```
Comment (ruchim) [inline bam-to-unmapped-bams.wdl]: v32?

File: bam-to-unmapped-bams.wdl [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+  }
+
+  scatter (unmapped_bam in RevertSam.unmapped_bams) {
+    String output_basename = basename(unmapped_bam, ".coord.sorted.unmapped.bam")
+    Float unmapped_bam_size = size(unmapped_bam, "GB")
+
+    call SortSam {
+      input:
+        input_bam = unmapped_bam,
+        sorted_bam_name = output_basename + ".unmapped.bam",
```
Comment (ruchim) [inline bam-to-unmapped-bams.wdl:66]: shouldn't need the `+ ".unmapped.bam"` anymore (as `output_basename` includes that extension)

File: paired-fastq-to-unmapped-bam-fc.wdl [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+  String gatk_path = select_first([gatk_path_override, "/gatk/gatk"])
+  Int? preemptible_attempts
+
+  # Convert multiple pairs of input fastqs in parallel
+  scatter (i in range(length(readgroup_array))) {
+
+    # Convert pair of FASTQs to uBAM
+    call PairedFastQsToUnmappedBAM {
+      input:
+      fastq_1 = readgroup_array[i][1],
```
Comment (ruchim) [inline paired-fastq-to-unmapped-bam-fc.wdl]: sorry this is a small thing, but it would be nice to indent this list of inputs :)

File: paired-fastq-to-unmapped-bam-fc.wdl [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+      run_date = readgroup_array[i][6],
+      platform_name = readgroup_array[i][7],
+      sequencing_center = readgroup_array[i][8],
+      gatk_path = gatk_path,
+      docker = gatk_docker,
+      preemptible_attempts = preemptible_attempts
+    }
+  }
+
+    #Create a file with a list of the generated ubams
```
Comment (ruchim) [inline paired-fastq-to-unmapped-bam-fc.wdl]: adjust spacing

File: paired-fastq-to-unmapped-bam-fc.wdl [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+  call CreateFoFN {
+    input:
+    array_of_files = PairedFastQsToUnmappedBAM.output_bam,
```
Comment (ruchim) [inline paired-fastq-to-unmapped-bam-fc.wdl]: indent

File: paired-fastq-to-unmapped-bam-fc.wdl [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+    --SAMPLE_NAME ${sample_name} \
+    --LIBRARY_NAME ${library_name} \
+    --PLATFORM_UNIT ${platform_unit} \
+    --RUN_DATE ${run_date} \
+    --PLATFORM ${platform_name} \
+    --SEQUENCING_CENTER ${sequencing_center} 
+  }
+  runtime {
+    docker: docker
+    memory: select_first([machine_mem_gb,10]) + " GB"
```
Comment (ruchim) [inline paired-fastq-to-unmapped-bam-fc.wdl]: space after comma

File: paired-fastq-to-unmapped-bam-fc.wdl [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+  String fofn_name
+  
+  # Runtime parameters
+  String docker
+  
+  command {
+    mv ${write_lines(array_of_files)}  ${fofn_name}.list
+  }
+  output {
+	  File fofn_list = "${fofn_name}.list"
```
Comment (ruchim) [inline paired-fastq-to-unmapped-bam-fc.wdl]: reduce indent

File: paired-fastq-to-unmapped-bam.inputs.json [hunk 14]
```diff
@@ -1,40 +1,30 @@
...
+  "##_Comment3": "GATK Luanch Path",
+  "#ConvertPairedFastQsToUnmappedBamWf.PairedFastQsToUnmappedBAM.gatk_path_override": "",
   
-  "##_Comment3": "Runtime Parameters",
-  "ConvertPairedFastQsToUnmappedBamWf.PairedFastQsToUnmappedBAM.docker": "broadinstitute/gatk:4.0.1.2",
-  "ConvertPairedFastQsToUnmappedBamWf.PairedFastQsToUnmappedBAM.mem_size": "1 GB",
-  "ConvertPairedFastQsToUnmappedBamWf.PairedFastQsToUnmappedBAM.disk_size": 200
-}
+  "##_Comment4": "Docker",
+  "#ConvertPairedFastQsToUnmappedBamWf.gatk_docker_override": "",
```
Comment (ruchim) [inline paired-fastq-to-unmapped-bam.inputs.json]: Similar to the line below, possibly replace empty `""` with `(optional) String?`

File: paired-fastq-to-unmapped-bam-fc.wdl [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+## licensing information pertaining to the included programs.
+
+# WORKFLOW DEFINITION
+workflow ConvertPairedFastQsToUnmappedBamWf {
+  File readgroup_list  
+  Array[Array[String]] readgroup_array = read_tsv(readgroup_list) 
+  String ubam_list_name = basename(readgroup_list,".list") + "unmapped.bam.list"
+  
+  String? gatk_docker_override
+  String gatk_docker = select_first([gatk_docker_override, "broadinstitute/gatk:latest"])
```
Comment (ruchim) [inline paired-fastq-to-unmapped-bam-fc.wdl]: can the default docker be the ones in gcr instead of dockerhub? Especially as this is the FireCloud version?

File: paired-fastq-to-unmapped-bam.wdl [hunk 16]
```diff
@@ -26,74 +28,126 @@
...
+      run_date = run_date[i],
+      platform_name = platform_name[i],
+      sequencing_center = sequencing_center[i],
+      gatk_path = gatk_path,
+      docker = gatk_docker,
+      preemptible_attempts = preemptible_attempts
     }
   }
 
+    #Create a file with a list of the generated ubams
```
Comment (ruchim) [inline paired-fastq-to-unmapped-bam.wdl]: spacing

File: paired-fastq-to-unmapped-bam.wdl [hunk 17]
```diff
@@ -26,74 +28,126 @@
...
+  call CreateFoFN {
+    input:
+    array_of_files = PairedFastQsToUnmappedBAM.output_bam,
```
Comment (ruchim) [inline paired-fastq-to-unmapped-bam.wdl]: indent

File: paired-fastq-to-unmapped-bam.wdl [hunk 18]
```diff
@@ -26,74 +28,126 @@
...
+    --LIBRARY_NAME ${library_name} \
+    --PLATFORM_UNIT ${platform_unit} \
+    --RUN_DATE ${run_date} \
+    --PLATFORM ${platform_name} \
+    --SEQUENCING_CENTER ${sequencing_center} 
   }
   runtime {
     docker: docker
-    memory: mem_size
+    memory: select_first([machine_mem_gb,10]) + " GB"
```
Comment (ruchim) [inline paired-fastq-to-unmapped-bam.wdl]: spacing after comma

**Review by ruchim — APPROVED** [review]
Most suggestions are around syntax. For the FC version of these WDLs, it would be good to have the default docker be from GCR and not DockerHub. Secondarily, I believe there's a place where the extensions need adjustment. Other than that, it looks good to me!

File: bam-to-unmapped-bams.wdl (hunk 8 above)
Comment (bshifaw) [reply bam-to-unmapped-bams.wdl:66]: I think `basename(unmapped_bam, ".coord.sorted.unmapped.bam")` removes the `unmapped.bam` part so this line adds back without the `.coord.sorted`

File: bam-to-unmapped-bams.wdl (hunk 7 above)
Comment (bshifaw) [reply bam-to-unmapped-bams.wdl]: done

File: paired-fastq-to-unmapped-bam-fc.wdl (hunk 15 above)
Comment (bshifaw) [reply paired-fastq-to-unmapped-bam-fc.wdl]: i've modified the ones in FC to use gcr. The idea was to have the ones in git use docker to avoid any authentication issues for users running on different platforms.

File: paired-fastq-to-unmapped-bam-fc.wdl (hunk 9 above)
Comment (bshifaw) [reply paired-fastq-to-unmapped-bam-fc.wdl]: done

File: paired-fastq-to-unmapped-bam-fc.wdl (hunk 10 above)
Comment (bshifaw) [reply paired-fastq-to-unmapped-bam-fc.wdl]: done

File: paired-fastq-to-unmapped-bam-fc.wdl (hunk 11 above)
Comment (bshifaw) [reply paired-fastq-to-unmapped-bam-fc.wdl]: done

File: paired-fastq-to-unmapped-bam-fc.wdl (hunk 12 above)
Comment (bshifaw) [reply paired-fastq-to-unmapped-bam-fc.wdl]: done

File: paired-fastq-to-unmapped-bam-fc.wdl (hunk 13 above)
Comment (bshifaw) [reply paired-fastq-to-unmapped-bam-fc.wdl]: done

File: paired-fastq-to-unmapped-bam.inputs.json (hunk 14 above)
Comment (bshifaw) [reply paired-fastq-to-unmapped-bam.inputs.json]: done

File: bam-to-unmapped-bams.wdl (hunk 8 above)
Comment (ruchim) [reply bam-to-unmapped-bams.wdl:66]: fudge--YES

File: paired-fastq-to-unmapped-bam-fc.wdl (hunk 15 above)
Comment (bshifaw) [reply paired-fastq-to-unmapped-bam-fc.wdl]: oops no i didn't , I forgot this was the FC version. thanks for catching this.

File: paired-fastq-to-unmapped-bam.wdl (hunk 16 above)
Comment (bshifaw) [reply paired-fastq-to-unmapped-bam.wdl]: done

File: paired-fastq-to-unmapped-bam.wdl (hunk 17 above)
Comment (bshifaw) [reply paired-fastq-to-unmapped-bam.wdl]: done

File: paired-fastq-to-unmapped-bam.wdl (hunk 18 above)
Comment (bshifaw) [reply paired-fastq-to-unmapped-bam.wdl]: done
